CHANGE/BUGFIX: validate label selectors to prevent reconciliation failures

slashpai opened the pull request:
## Description

_Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request._

_If it fixes an existing issue (bug or feature), use the following keyword:_

Apply the 63 character limit to the name segment only and enforce a 253 character prefix limit per Kubernetes label syntax. Add e2e coverage for accepted and rejected prefixed keys.

Related-to: #7915


Follow-up to be done for Alertmanager, ThanosRuler, Config CRDs

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [x] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [ ] `FEATURE` (non-breaking change which adds functionality)
- [x] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Verification
<!-- How you tested it? How do you know it works? -->
Please check the [Prometheus-Operator testing guidelines](../TESTING.md) for recommendations about automated tests.

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

<!-- 
Your release note should be written in clear and straightforward sentences. Most often, users aren't familiar with
the technical details of your PR, so consider what they need to know when you write your release note.

Some brief examples of release notes:
- Add metadataConfig field to the Prometheus CRD for configuring how remote-write sends metadata information.
- Generate correct scraping configuration for Probes with empty or unset module parameter.
-->

```release-note
CHANGE/BUGFIX: Reject invalid label selectors on Prometheus and PrometheusAgent at the
API level using CEL validation for label key length and matchExpressions
operators.
```

simonpasquier reviewed (commented):
should we create our own type (copied from metav1.LabelSelector) to avoid repeating the validation markers?

slashpai committed:
BUGFIX: fix ValidatedLabelSelector length checks for prefixed label keys

Apply the 63 character limit to the name segment only and enforce a 253
character prefix limit per Kubernetes label syntax. Add e2e coverage for
accepted and rejected prefixed keys.

Related-to: #7915

Signed-off-by: Jayapriya Pai <slashpai9@gmail.com>

slashpai commented:
@simonpasquier Can you take a look at this when you get chance?

simonpasquier commented on pkg/apis/monitoring/v1/types.go:
```diff
@@ -537,6 +538,30 @@ func (c *WebTLSConfig) Validate() error {
+// +kubebuilder:validation:XValidation:rule="!has(self.matchLabels) || self.matchLabels.all(key, size(key) <= 63)",message="matchLabels keys must not exceed 63 characters"
+// +kubebuilder:validation:XValidation:rule="!has(self.matchExpressions) || self.matchExpressions.all(expr, size(expr.key) <= 63)",message="matchExpressions keys must not exceed 63 characters"
+// +kubebuilder:validation:XValidation:rule="!has(self.matchExpressions) || self.matchExpressions.all(expr, expr.operator in ['In', 'NotIn', 'Exists', 'DoesNotExist'])",message="matchExpressions operator must be one of: In, NotIn, Exists, DoesNotExist"
+type ValidatedLabelSelector struct {
```
(nit) can we rename LabelSelector?

slashpai replied:
I thought this would make it clear it's validated than regular one. But no strong opinion, ok to change if needed

simonpasquier reviewed (approved):
LGTM, just one nit.

simonpasquier reviewed (changes requested):
Ah after researching more 63 chars is the hard limit when there's no prefix but the total limit is higher if there's a prefix:
https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set
